openfisca_core/periods.py

```diff
@@ -751,135 +727,90 @@ def instant_date(instant):
-    >>> period('day', u'2014-02')
-    Period((u'day', Instant((2014, 2, 1)), 28))
-
-    >>> period('year', u'2014-3-2')
-    Period((u'year', Instant((2014, 3, 2)), 1))
-    >>> period('month', u'2014-3-2')
-    Period((u'month', Instant((2014, 3, 2)), 1))
-    >>> period('day', u'2014-3-2')
-    Period((u'day', Instant((2014, 3, 2)), 1))
-
-    >>> period('year', u'2014-3-2', size = 2)
-    Period((u'year', Instant((2014, 3, 2)), 2))
-    >>> period('month', u'2014-3-2', size = 2)
-    Period((u'month', Instant((2014, 3, 2)), 2))
-    >>> period('day', u'2014-3-2', size = 2)
-    Period((u'day', Instant((2014, 3, 2)), 2))
-
-    >>> period('month', instant(u'2014-3-2'), size = 2)
-    Period((u'month', Instant((2014, 3, 2)), 2))
-    >>> period('month', period(u'year', u'2014-3-2'), size = 2)
-    Period((u'month', Instant((2014, 3, 2)), 2))
+    Period((MONTH, Instant((2014, 2, 1)), 1))
+
+    >>> period(u'year:2014-2')
+    Period((YEAR, Instant((2014, 2, 1)), 1))
     """
-    if not isinstance(value, basestring) or value not in (u'day', u'month', u'year'):
-        assert start is None, start
-        assert size is None, size
-        return conv.check(json_or_python_to_period)(value)
-    unit = unicode(value)
-    assert size is None or isinstance(size, int) and size > 0, size
-
-    if isinstance(start, basestring):
-        start = tuple(
-            int(fragment)
-            for fragment in start.split(u'-', 2)[:3]
-            )
-    elif isinstance(start, datetime.date):
-        start = (start.year, start.month, start.day)
-    elif isinstance(start, int):
-        start = (start,)
-    elif isinstance(start, list):
-        assert 1 <= len(start) <= 3
-        start = tuple(start)
-    elif isinstance(start, Period):
-        start = start.start
-    else:
-        assert isinstance(start, tuple)
-        assert 1 <= len(start) <= 3
-    if len(start) == 1:
-        start = Instant((start[0], 1, 1))
-        if size is None:
-            if unit == u'day':
-                size = 366 if calendar.isleap(start[0]) else 365
-            elif unit == u'month':
-                size = 12
-            else:
-                size = 1
-    elif len(start) == 2:
-        start = Instant((start[0], start[1], 1))
-        if size is None:
-            if unit == u'day':
-                size = calendar.monthrange(start[0], start[1])[1]
-            else:
-                size = 1
+
+    def parse_iso_period(value):
+        """
+        Parses simple periods respecting the ISO format, such as 2012 or 2015-03
+        """
+        try:
+            date = datetime.datetime.strptime(value, '%Y')
+            return Period((YEAR, Instant((date.year, date.month, 1)), 1))
+        except ValueError:
+            pass
+        try:
+            date = datetime.datetime.strptime(value, '%Y-%m')
+            return Period((MONTH, Instant((date.year, date.month, 1)), 1))
+        except ValueError:
+            return None
```

Comment:
I would prefer:

```python
try:
    date = datetime.datetime.strptime(value, '%Y')
except ValueError:
    try:
        date = datetime.datetime.strptime(value, '%Y-%m')
    except ValueError:
        return None
    else:
        return Period((MONTH, Instant((date.year, date.month, 1)), 1))
else:
    return Period((YEAR, Instant((date.year, date.month, 1)), 1))
```

It's less readable, but it wraps only the `strptime` calls in a `try/except`